[DDP] log gradient ready order and bucket indices

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62770
* __->__ #62751
* #62748

This will help us determine whether gradient ready order and bucket
indices are aligned amongst all the ranks. This should always be true for rank
0 as we determine rebuilt bucket order by the gradient ready order on rank 0,
but would be interested to see this on different workloads for other ranks

Differential Revision: [D30111833](https://our.internmc.facebook.com/intern/diff/D30111833/)

rohan-varma committed:
[DDP] log gradient ready order and bucket indices

This will help us determine whether gradient ready order and bucket
indices are aligned amongst all the ranks. This should always be true for rank
0 as we determine rebuilt bucket order by the gradient ready order on rank 0,
but would be interested to see this on different workloads for other ranks

Differential Revision: [D30111833](https://our.internmc.facebook.com/intern/diff/D30111833/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] log gradient ready order and bucket indices"


This will help us determine whether gradient ready order and bucket
indices are aligned amongst all the ranks. This should always be true for rank
0 as we determine rebuilt bucket order by the gradient ready order on rank 0,
but would be interested to see this on different workloads for other ranks

Differential Revision: [D30111833](https://our.internmc.facebook.com/intern/diff/D30111833/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] log gradient ready order and bucket indices"


This will help us determine whether gradient ready order and bucket
indices are aligned amongst all the ranks. This should always be true for rank
0 as we determine rebuilt bucket order by the gradient ready order on rank 0,
but would be interested to see this on different workloads for other ranks

Differential Revision: [D30111833](https://our.internmc.facebook.com/intern/diff/D30111833/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] log gradient ready order and bucket indices"


This will help us determine whether gradient ready order and bucket
indices are aligned amongst all the ranks. This should always be true for rank
0 as we determine rebuilt bucket order by the gradient ready order on rank 0,
but would be interested to see this on different workloads for other ranks

Differential Revision: [D30111833](https://our.internmc.facebook.com/intern/diff/D30111833/)

[ghstack-poisoned]

zhaojuanmao commented on torch/csrc/distributed/c10d/logger.cpp:
```diff
@@ -237,6 +248,23 @@ void Logger::set_runtime_stats_and_log() {
+  for (const auto& bucket_indices : indices) {
+    per_bucket_variable_indices.push_back(c10::Join(" ", bucket_indices));
+  }
+  ddp_logging_data_->strs_map["per_bucket_variable_indices"] =
```
nit: rebuilt_per_bucket_param_indices

zhaojuanmao commented on torch/csrc/distributed/c10d/logger.cpp:
```diff
@@ -237,6 +248,23 @@ void Logger::set_runtime_stats_and_log() {
+        c10::Join(", ", reducer_->grad_ready_order_indices_);
+  }
+  // Log per-bucket variable indices
+  std::vector<std::string> per_bucket_variable_indices;
```
move this logging to line 249, so that it is only be called once

zhaojuanmao reviewed (approved):
lgtm, some nits commented

rohan-varma committed:
Update on "[DDP] log gradient ready order and bucket indices"


This will help us determine whether gradient ready order and bucket
indices are aligned amongst all the ranks. This should always be true for rank
0 as we determine rebuilt bucket order by the gradient ready order on rank 0,
but would be interested to see this on different workloads for other ranks

Differential Revision: [D30111833](https://our.internmc.facebook.com/intern/diff/D30111833/)

[ghstack-poisoned]
